tokenizers/src/pre_tokenizers/split.rs

```diff
@@ -0,0 +1,145 @@
+use std::fmt;
+
+use regex::Regex;
+use serde::{Deserialize, Deserializer, Serialize};
+
+// use crate::pattern::Pattern;
+use crate::tokenizer::{
+    pattern::{Invert, Pattern}, PreTokenizedString, PreTokenizer, Result, SplitDelimiterBehavior,
+};
+use serde::de::{Error, Visitor};
+
+#[derive(Clone, Debug, Serialize)]
+#[serde(tag = "type")]
+pub struct Split {
+    #[serde(default = "default_regex", skip)]
+    re: Regex,
+    #[serde(default = "default_behavior", skip)]
+    behavior: SplitDelimiterBehavior,
+    invert: bool
+}
+
+fn default_regex() -> Regex {
+    Regex::new(r"\w+|[^\w\s]+").unwrap()
+}
+
+fn default_behavior() -> SplitDelimiterBehavior {
+    SplitDelimiterBehavior::Removed
+}
+
+impl Default for Split {
+    fn default() -> Self {
+        Self {
+            re: default_regex(),
+            behavior: default_behavior(),
+            invert: false,
+        }
+    }
+}
+
+impl Split {
+    pub fn new(re: Regex, behavior: SplitDelimiterBehavior, invert: bool) -> Self {
+        Split {
+            re,
+            behavior,
+            invert
+        }
+    }
+}
+
+impl PreTokenizer for Split {
+    fn pre_tokenize(&self, pretokenized: &mut PreTokenizedString) -> Result<()> {
+        use SplitDelimiterBehavior::*;
+        match self.behavior {
+            Removed => {
```

Comment:
this is far from being clean. I encountered problems with passing the correct `SplitDelimiterBehavior` without having to do a match. 

```
pretokenized.split(|_, normalized| {
                        normalized.split(Invert(&self.re), self.behavior)
})
```

throws a "move" error. Would be super happy for some pointers to resources on how this can be better handled!!!

Reply:
Actually, this is one of those things Rust handles very well. You can't make a type `Copy` if it contains anything that shouldn't be copied, like `Vec`, `HashMap`, ...

So, as the Rust doc says:
> **When should my type be Copy?**
Generally speaking, if your type can implement Copy, it should. Keep in mind, though, that implementing Copy is part of the public API of your type. If the type might become non-Copy in the future, it could be prudent to omit the Copy implementation now, to avoid a breaking API change.

(source [the doc for `std::marker::Copy`](https://doc.rust-lang.org/std/marker/trait.Copy.html))